packages/contracts/evm-contracts/contracts/orderbook/OrderbookDex.sol

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+import {ERC165} from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import {IOrderbookDex} from "./IOrderbookDex.sol";
+
+/// @notice Facilitates trading an asset that is living on a different app-chain.
+contract OrderbookDex is IOrderbookDex, ERC165, ReentrancyGuard {
+    using Address for address payable;
+
+    mapping(uint256 => Order) public orders;
+    uint256 public ordersIndex;
+
+    error OrderIsInactive(uint256 orderId);
+    error Unauthorized();
+
+    /// @notice Returns the current index of orders (index that a new sell order will be mapped to).
+    function getOrdersIndex() public view returns (uint256) {
+        return ordersIndex;
+    }
+
+    /// @notice Returns the Order struct information about order of specified `orderId`.
+    function getOrder(uint256 orderId) public view returns (Order memory) {
+        return orders[orderId];
+    }
+
+    /// @notice Creates a sell order for the specified `assetAmount` at specified `price`.
+    /// @dev The order is saved in a mapping from incremental ID to Order struct.
+    function createSellOrder(uint256 assetAmount, uint256 price) public {
+        Order memory newOrder = Order({
+            seller: payable(msg.sender),
+            assetAmount: assetAmount,
+            price: price,
+            active: true
+        });
+        orders[ordersIndex] = newOrder;
+        emit OrderCreated(ordersIndex, msg.sender, assetAmount, price);
+        ordersIndex++;
+    }
+
+    /// @notice Fills an array of orders specified by `orderIds`.
+    /// @dev Reverts if msg.value is less than the sum of orders' prices.
+    /// If msg.value is more than the sum of orders' prices, it should refund the difference back to msg.sender.
+    function fillSellOrders(uint256[] memory orderIds) public payable nonReentrant {
+        uint256 length = orderIds.length;
+        uint256 totalPaid;
+        for (uint256 i = 0; i < length; ) {
+            uint256 orderId = orderIds[i];
+            Order memory order = orders[orderId];
+            if (!order.active) {
+                revert OrderIsInactive(orderId);
+            }
+            order.seller.sendValue(order.price);
+            totalPaid += order.price;
+            orders[orderId].active = false;
+            emit OrderFilled(orderId, order.seller, msg.sender, order.assetAmount, order.price);
+            unchecked {
```

Comment:
this function needs to also create a new order if an order was partially filled.

Ex: if an order is to sell 5000 TGOLD
And you buy 3000 TGOLD

It should mark the 5000 order as inactive (to avoid concurrency issues) and create a new order with 2000 tgold

Reply:
I didn't know we wanted to support partial fills. In that case:
1. Shouldn't we make it so that the seller can choose if they want their sell order to be single fill or if they want to allow partial fills? (afaik this is how it's done on whales.market for example)
2. In the case of partial fill, why would we have to discard the old order and create a new order? I understand that you're thinking in terms of UTxO model, but is this really necessary? Example: There is sell order for 1000 TGOLD. I submit a buy for 800 TGOLD but somebody else submits a buy of 600 TGOLD with higher gas fee, so by the time my transaction is being processed, 600 was already bought and so the sell order would only have 400 TGOLD left. I would I would much rather have my buy order fill for the 400 TGOLD (because it was a good price per token) than it to revert and me not getting anything and losing money on failed transaction gas fee. To be extra inclusive, we can make it an option for the buyer - "allow partial buy fill" (because maybe somebody really does want their buy order to either fill entirely or fail).

Reply:
I moved my comment on this topic here: https://github.com/PaimaStudios/PRC/pull/5#discussion_r1525212545

Reply:
> Shouldn't we make it so that the seller can choose if they want their sell order to be single fill or if they want to allow partial fills? (afaik this is how it's done on whales.market for example)

I don't think this is necessary. The reason whales.market does this is because users have to manually choose which order they want to fill, but we don't have that issue since it's all automated by the UI. At most we could do something to avoid dust (ex: an entry having so little tgold in it that it's not worth the tx fee to include it), but that's a little complicated